privacyidea.py

```diff
@@ -2,131 +2,241 @@
+                        
+                        # Check if an input field is needed for any other token type
+                        if tttList.size() > 0:
+                            identity.setWorkingParameter("otp_available", "1")
+
+                        return True
+                else:
+                    print("privacyIDEA. Empty response from server")
+                    fm.add(FacesMessage.SEVERITY_ERROR, "No response from the privacyIDEA Server. Please check the connection!")
+
+                return False
+            else:
+                #print("privacyIDEA. Username is empty")
+                fm.add(FacesMessage.SEVERITY_ERROR, "Please enter a username!")
 
-    def prepareForStep(self, configurationAttributes, requestParameters, step):
-        if step == 1:
-            print "privacyIDEA. Prepare for Step 1"
-            return True
-        elif step == 2:
-            # This is a 2nd challenge response step
-            print("privacyIDEA. Prepare for another Step {0!s}.".format(step))
-            return True
-        else:
             return False
+        else:
+            # Get the user from step 1
+            currentUser = self.getFromSession("currentUser")
+
+            if currentUser:
+                authenticationService.authenticate(currentUser)
+            else:
+                print("privacyIDEA. No user found in session for second step")
+                fm.add(FacesMessage.SEVERITY_ERROR, "An error occurred. Please try to restart the authentication!")
+                return False
+
+            try:
+                # Persist the mode between the script and the js
+                mode = requestParameters.get("modeField")[0].strip()
+                identity.setWorkingParameter("mode", mode)
+            except TypeError:
+                print("privacyIDEA. Mode not found in request parameters")
+
+            txid = self.getFromSession("transaction_id")
+
+            # If mode is push: poll for the transactionID to see if the user confirmed on the smartphone
+            if mode == "push":
+                if not txid:
+                    print("privacyIDEA. Transaction ID not found in session, but it is mandatory for polling!")
+                    fm.add(FacesMessage.SEVERITY_ERROR, "Your transaction id could not be found. Please try to restart the authentication!")
+                    return False
+
+                if self.pi.pollTransaction(txid):
+                    # If polling is successful, the authentication has to be finished by a call to validateCheck
+                    # with the username, NO otp and the transactionID
+                    response = self.pi.validateCheck(currentUser, "", txid)
+                    return response.getValue()
+
+            elif mode == "otp":
+                try:
+                    otp = requestParameters.get("otp")[0].strip()
+                    #print("OTP: %s" % otp)
+                except TypeError:
```

Comment:
When would you get a TypeError? Is requestParameters.get("otp") supposed to be a list?

Reply:
yes